onnx/src/ops/ml/mod.rs

```diff
@@ -0,0 +1,330 @@
+            Node::new_branch(cmp, feature_id, value, true_id, false_id, nan_is_true)
+        } else {
+            if let Some(&(start, end)) = leaf_map.get(&(tree_id, node_id)) {
+                Node::new_leaf(start, end)
+            } else {
+                node.bail(&format!("leaf not found: tree_id={}, node_id={}", tree_id, node_id))?
+            }
+        };
+        nodes[tree_id].push(tree_node);
+
+        prev_tree_id = tree_id;
+        prev_node_id = node_id;
+    }
+
+    // build the trees
+
+    let mut trees: Vec<Tree> = Vec::default();
+
+    for i in 0..n_trees {
+        let tree = Tree::build(n_classes, &nodes[i], &leaves[i])
+            .or_else(|err| node.bail(&format!("{}", err)))?;
+        trees.push(tree);
+    }
+
+    // build the tree ensemble
+
+    let base_scores = base_values.as_ref().map(Vec::as_slice);
+    let ensemble = TreeEnsemble::build(&trees, aggregate_fn, post_transform, base_scores)
+        .or_else(|err| node.bail(&format!("{}", err)))?;
+
+    Ok(ensemble)
+}
+
+pub fn register_all_ops(reg: &mut OpRegister) {
+    reg.insert("TreeEnsembleClassifier", TreeEnsembleClassifier::parse);
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ClassLabels {
+    Ints(Vec<i64>),
+    Strings(Vec<String>),
+}
+
+#[derive(Debug, Clone)]
+pub struct TreeEnsembleClassifier {
+    ensemble: TreeEnsemble,
+    class_labels: ClassLabels,
+}
+
+impl TreeEnsembleClassifier {
+    fn parse(node: &NodeProto) -> TractResult<Box<Op>> {
+        let ensemble = parse_tree_ensemble(node, true)?;
+        let class_labels = match node.get_attr_opt_slice::<i64>("classlabels_int64s")? {
+            Some(int_labels) => ClassLabels::Ints(int_labels.into()),
+            _ => {
+                let str_labels = node.get_attr_opt_vec::<String>("classlabels_strings")?;
+                ClassLabels::Strings(str_labels.unwrap())
+            }
+        };
+        Ok(Box::new(Self { ensemble, class_labels }))
+    }
+}
+
+impl Op for TreeEnsembleClassifier {
+    fn name(&self) -> Cow<str> {
+        "TreeEnsembleClassifier".into()
+    }
+
+    fn noutputs(&self) -> usize {
+        2
+    }
+}
+
+impl StatelessOp for TreeEnsembleClassifier {
+    fn eval(&self, mut inputs: TVec<SharedTensor>) -> TractResult<TVec<SharedTensor>> {
+        Err("WIP".into())
+    }
+}
+
+impl InferenceRulesOp for TreeEnsembleClassifier {
+    fn rules<'r, 'p: 'r, 's: 'r>(
+        &'s self, s: &mut Solver<'r>, inputs: &'p [TensorProxy], outputs: &'p [TensorProxy],
+    ) -> InferenceResult {
+        check_input_arity(&inputs, 1)?;
+        check_output_arity(&outputs, 2)?;
+
+        s.equals(&inputs[0].rank, 2)?;
+        s.equals(&outputs[0].rank, 1)?;
+        s.equals(&outputs[0].rank, 1)?;
+
+        s.given_3(
+            &inputs[0].shape,
+            &outputs[0].shape,
+            &outputs[1].shape,
+            move |s, s_in0, s_out0, s_out1| {
+                s.equals(s_in0[0], s_out0[0])?;
```

Comment:
unless I'm missing something this two shape constraint don't need to be in the given block. The benefit of being outside of the given block is that it will allow to make partial deduction on shape individual values, while the given() blocks are only executed if and when all the arguments are determined.

```
s.equals(&inputs[0].shape[0], &outputs[0].shape[0])?;
```

should do the trick, right after the rank checks (order does not matters).

Reply:
Makes sense, thanks! I'll change it to simple equalities. I looked at the existing code but wasn't sure when exactly you're supposed to use given().